// deck.gl/dist/lib/utils/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.log = exports.checkArray = exports.compareArrays = exports.fillArray = exports.flattenVertices = exports.flatten = exports.get = exports.count = exports.Container = exports.parseColor = undefined;

var _color = require('./color');

Object.defineProperty(exports, 'parseColor', {
  enumerable: true,
  get: function get() {
    return _color.parseColor;
  }
});

var _container = require('./container');

Object.defineProperty(exports, 'count', {
  enumerable: true,
  get: function get() {
    return _container.count;
  }
});
Object.defineProperty(exports, 'get', {
  enumerable: true,
  get: function get() {
    return _container.get;
  }
});

var _flatten = require('./flatten');

Object.defineProperty(exports, 'flatten', {
  enumerable: true,
  get: function get() {
    return _flatten.flatten;
  }
});
Object.defineProperty(exports, 'flattenVertices', {
  enumerable: true,
  get: function get() {
    return _flatten.flattenVertices;
  }
});
Object.defineProperty(exports, 'fillArray', {
  enumerable: true,
  get: function get() {
    return _flatten.fillArray;
  }
});

var _compareObjects = require('./compare-objects');

Object.keys(_compareObjects).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _compareObjects[key];
    }
  });
});

var _compareArrays = require('./compare-arrays');

Object.defineProperty(exports, 'compareArrays', {
  enumerable: true,
  get: function get() {
    return _compareArrays.compareArrays;
  }
});
Object.defineProperty(exports, 'checkArray', {
  enumerable: true,
  get: function get() {
    return _compareArrays.checkArray;
  }
});

var _log = require('./log');

Object.defineProperty(exports, 'log', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_log).default;
  }
});

var _fp = require('./fp64');

Object.keys(_fp).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _fp[key];
    }
  });
});

var _blend = require('./blend');

Object.keys(_blend).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _blend[key];
    }
  });
});

var Container = _interopRequireWildcard(_container);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.Container = Container;
