// luma.gl/dist/webgl/sync.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }(); // WebGL2 Sync Object Helper
// https://developer.mozilla.org/en-US/docs/Web/API/WebGLQuery


var _webglChecks = require('../webgl/webgl-checks');

var _queryManager = require('./queryManager');

var _queryManager2 = _interopRequireDefault(_queryManager);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

// WebGLSync? fenceSync(GLenum condition, GLbitfield flags);
// [WebGLHandlesContextLoss] GLboolean isSync(WebGLSync? sync);
// void deleteSync(WebGLSync? sync);
// GLenum clientWaitSync(WebGLSync? sync, GLbitfield flags, GLint64 timeout);
// void waitSync(WebGLSync? sync, GLbitfield flags, GLint64 timeout);
// any getSyncParameter(WebGLSync? sync, GLenum pname);

var Sync = function () {
  /**
   * @class
   * @param {WebGL2RenderingContext} gl
   */
  function Sync(gl) {
    _classCallCheck(this, Sync);

    (0, _webglChecks.assertWebGL2Context)(gl);

    var handle = gl.fenceSync(gl.SYNC_GPU_COMMANDS_COMPLETE, 0);
    (0, _webglChecks.glCheckError)(gl);

    this.gl = gl;
    this.handle = handle;
    this.userData = {};

    // query manager needs a promise field
    this.promise = null;

    Object.seal(this);
  }

  /**
   * @return {Sync} returns self to enable chaining
   */


  _createClass(Sync, [{
    key: 'delete',
    value: function _delete() {
      _queryManager2.default.deleteQuery(this);
      if (this.handle) {
        this.gl.deleteSync(this.handle);
        this.handle = null;
        (0, _webglChecks.glCheckError)(this.gl);
      }
      return this;
    }

    /**
     * @param {GLbitfield} flags
     * @param {GLint64} timeout
     * @return {Sync} returns self to enable chaining
     */

  }, {
    key: 'wait',
    value: function wait(flags, timeout) {
      this.gl.waitSync(this.handle, flags, timeout);
      (0, _webglChecks.glCheckError)(this.gl);
      return this;
    }

    /**
     * @param {GLbitfield} flags
     * @param {GLint64} timeout
     * @return {GLenum} result
     */

  }, {
    key: 'clientWait',
    value: function clientWait(flags, timeout) {
      var result = this.gl.clientWaitSync(this.handle, flags, timeout);
      (0, _webglChecks.glCheckError)(this.gl);
      return result;
    }
  }, {
    key: 'cancel',
    value: function cancel() {
      _queryManager2.default.cancelQuery(this);
    }
  }, {
    key: 'isResultAvailable',
    value: function isResultAvailable() {
      var status = this.gl.getSyncParameter(this.handle, this.gl.SYNC_STATUS);
      return status === this.gl.SIGNALED;
    }
  }, {
    key: 'getResult',
    value: function getResult() {
      return this.gl.SIGNALED;
    }
  }]);

  return Sync;
}();

exports.default = Sync;
