// luma.gl/dist/experimental/probe/src/env.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.timestamp = exports.logger = exports.IS_NODE = undefined;

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }();

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; /**
                                                                                                                                                                                                                                                                               * Common environment setup
                                                                                                                                                                                                                                                                               */
/* eslint-disable no-console */
/* global process */


var _console = require('global/console');

var _console2 = _interopRequireDefault(_console);

var _window = require('global/window');

var _window2 = _interopRequireDefault(_window);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Duck-type Node context
var IS_NODE = exports.IS_NODE = (typeof process === 'undefined' ? 'undefined' : _typeof(process)) !== undefined && process.toString() === '[object process]';

// Configure console

// Console.debug is useful in chrome as it gives blue styling, but is not
// available in node
_console2.default.debug = _console2.default.debug || _console2.default.log;

// Some instrumentation may override console methods, so preserve them here
_console2.default.native = {
  debug: _console2.default.debug.bind(_console2.default),
  log: _console2.default.log.bind(_console2.default),
  warn: _console2.default.warn.bind(_console2.default),
  error: _console2.default.error.bind(_console2.default)
};

exports.logger = _console2.default;

// Set up high resolution timer

var timestamp = void 0;
if (IS_NODE) {
  exports.timestamp = timestamp = function timestamp() {
    var _process$hrtime = process.hrtime(),
        _process$hrtime2 = _slicedToArray(_process$hrtime, 2),
        seconds = _process$hrtime2[0],
        nanoseconds = _process$hrtime2[1];

    return seconds + nanoseconds / 1e6;
  };
} else if (_window2.default.performance) {
  exports.timestamp = timestamp = function timestamp() {
    return _window2.default.performance.now();
  };
} else {
  exports.timestamp = timestamp = function timestamp() {
    return Date.now();
  };
}

exports.timestamp = timestamp;
