// react-map-gl/dist/overlays/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _canvas = require('./canvas.react');

Object.defineProperty(exports, 'CanvasOverlay', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_canvas).default;
  }
});

var _choropleth = require('./choropleth.react');

Object.defineProperty(exports, 'ChoroplethOverlay', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_choropleth).default;
  }
});

var _draggablePoints = require('./draggable-points.react');

Object.defineProperty(exports, 'DraggablePointsOverlay', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_draggablePoints).default;
  }
});

var _html = require('./html.react');

Object.defineProperty(exports, 'HTMLOverlay', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_html).default;
  }
});

var _scatterplot = require('./scatterplot.react');

Object.defineProperty(exports, 'ScatterplotOverlay', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_scatterplot).default;
  }
});

var _svg = require('./svg.react');

Object.defineProperty(exports, 'SVGOverlay', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_svg).default;
  }
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
